localstack/http/router.py

```diff
@@ -462,23 +373,223 @@ def route(
+        :param obj: the object to scan
         """
-        DEPRECATED: use ``add`` instead.
+        ...
+
+    @overload
+    def __init__(self, rules: list[Union[_RouteEndpoint, RuleFactory, Any]]):
+        """Add multiple rules at once"""
+        ...
+
+    def __init__(self, *args, **kwargs):
         """
-        return self._add_routes(obj)
+        Dispatcher for overloaded ``__init__`` methods.
+        """
+        if "path" in kwargs or type(args[0]) is str:
+            self.factory = _EndpointRule(*args, **kwargs)
+        elif "fn" in kwargs or callable(args[0]):
+            self.factory = _EndpointFunction(*args, **kwargs)
+        elif "rule_factory" in kwargs:
+            self.factory = kwargs["rule_factory"]
+        elif isinstance(args[0], RuleFactory):
+            self.factory = args[0]
+        elif isinstance(args[0], list):
+            self.factory = RuleGroup([RuleAdapter(rule) for rule in args[0]])
+        else:
+            self.factory = _EndpointsObject(*args, **kwargs)
+
+    def get_rules(self, map: Map) -> t.Iterable[Rule]:
+        yield from self.factory.get_rules(map)
+
+
+class WithHost(RuleFactory):
+    def __init__(self, host: str, rules: Iterable[RuleFactory]) -> None:
+        self.host = host
+        self.rules = rules
+
+    def get_rules(self, map: Map) -> t.Iterator[Rule]:
+        for rulefactory in self.rules:
+            for rule in rulefactory.get_rules(map):
+                rule = rule.empty()
+                rule.host = self.host
+                yield rule
+
+
+class RuleGroup(RuleFactory):
+    def __init__(self, rules: Iterable[RuleFactory]):
+        self.rules = rules
+
+    def get_rules(self, map: Map) -> t.Iterable[Rule]:
+        for rule in self.rules:
+            yield from rule.get_rules(map)
+
 
-    def remove_rule(self, rule: Rule):
-        """DEPRECATED: use ``remove`` instead."""
-        self._remove_rules([rule])
+class _EndpointRule(RuleFactory, Generic[E]):
+    def __init__(
+        self,
+        path: str,
+        endpoint: E,
+        host: Optional[str] = None,
+        methods: Optional[Iterable[str]] = None,
+        **kwargs,
+    ):
+        self.path = path
+        self.endpoint = endpoint
+        self.host = host
+        self.methods = methods
+        self.kwargs = kwargs
+
+    def get_rules(self, map: Map) -> t.Iterable[Rule]:
+        host = self.host
+
+        if host is None and map.host_matching:
+            # this creates a "match any" rule, and will put the value of the host
+            # into the variable "__host__"
+            host = "<__host__>"
+
+        # the typing for endpoint is a str, but the doc states it can be any value,
+        # however then the redirection URL building will not work
+        rule = Rule(
+            self.path, endpoint=self.endpoint, methods=self.methods, host=host, **self.kwargs
+        )
+        yield rule
+
+
+class _EndpointFunction(RuleFactory):
```

Comment:
Maybe you could add a small docstring explaining that this class is used for callables which have been decorated with the route decorator (which will take care of setting the expected `rule_attributes` as a member of the callable object)?